### PR TITLE
Unlimited SWAP and DUP instructions

### PR DESCRIPTION
**Pull request by axic**

### CODE REVIEW

**Comment by axic** [comment]
Alternatively to ease verification, the parameter to `SWAPn` / `DUPn` could be:
- force that a `PUSH` is present before them
- have an immediate value (2 bytes, to accommodate the value 1024) instead of a PUSH

**Comment by axic** [comment]
To clarify, the "immediate value" option would mean an encoding `0xb0 0xNN 0xNN` for DUP and `0bx1 0xNN 0xNN` for SWAP, just like how PUSH works.

**Comment by axic** [comment]
Sorry haven't seen the issue #174 which proposes the same.

**Comment by gcolvin** [comment]
I still think that if https://github.com/ethereum/EIPs/issues/615 is accepted this will be unnecessary.

File: EIPS/draft_unlimited_swapdup.md [hunk 1]
```diff
@@ -0,0 +1,57 @@
+## Preamble
+
+    EIP: <to be assigned>
```
Comment (pirapira) [inline EIPS/draft_unlimited_swapdup.md]: Please use 663 and move this file to `EIPS/eip-663.md`.

File: EIPS/draft_unlimited_swapdup.md [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+The number of these arguments (or stack items) can easily exceed 16 and thus will require extra care from a compiler to lay them out in a way that all of them are still accessible.
+
+Introducing `SWAPn` and `DUPn` will provide an option to compilers to simplify accessing deep stack items at the price of possibly increased gas costs.
+
+## Specification
+
+Instructions `DUPn` (`0xb0`) and `SWAPn` (`0xb1`) are introduced, which take the top item from stack (referred to as `n`).
+
+If `n` exceeds 1024 or the current stack depth is less than `n`, then a stack underflow exception is issued. If the current stack depth is at the limit, a stack overflow exception is issued.
```
Comment (pirapira) [inline EIPS/draft_unlimited_swapdup.md]: When a stack underflow exception is issued, what should happen to the contract execution?  Would it be equivalent to `REVERT` with a certain output?

Does it ever happen that the stack depth is still at the limit after popping `n`?

File: EIPS/draft_unlimited_swapdup.md [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+Otherwise
+- for `DUPn` the stack item at depth `n` is duplicated at the top of the stack
+- for `SWAPn` the top stack item is swapped with the item at depth `n`
+
+The gas cost for both instructions is set at 3. In reality the cost for such an operation is 6 including the required `PUSH`.
```
Comment (pirapira) [inline EIPS/draft_unlimited_swapdup.md]: The cost can be smaller than 6 per operation.  In fact the cost can be very close to 3.  For example,
```
PUSH1 0
PUSH1 0
DUPn
DUPn
DUPn
DUPn
DUPn
DUPn
DUPn
...
```

**Comment by Arachnid** [comment]
This is a courtesy notice to let you know that the format for EIPs has been modified slightly. If you want your draft merged, you will need to make some small changes to how your EIP is formatted:

 - Frontmatter is now contained between lines with only a triple dash ('---')
 - Headers in the frontmatter are now lowercase.

If your PR is editing an existing EIP rather than creating a new one, this has already been done for you, and you need only rebase your PR.

In addition, a continuous build has been setup, which will check your PR against the rules for EIP formatting automatically once you update your PR. This build ensures all required headers are present, as well as performing a number of other checks.

Please rebase your PR against the latest master, and edit your PR to use the above format for frontmatter. For convenience, here's a sample header you can copy and adapt:

```
---
eip: <num>
title: <title>
author: <author>
type: [Standards Track|Informational|Meta]
category: [Core|Networking|Interface|ERC] (for type: Standards Track only)
status: Draft
created: <date>
---
```

**Comment by axic** [comment]
Updated to satisfy draft requirements.

**Comment by MrChico** [comment]
It seems a lot easier to do stack analysis if arguments inlined in the code, rather than taken off the stack. That would also play better with #615 .

**Comment by axic** [comment]
> It seems a lot easier to do stack analysis if the would be arguments inlined in the code, rather than taken off the stack.

This option was also proposed, see https://github.com/ethereum/EIPs/pull/663#issuecomment-312950394 and https://github.com/ethereum/EIPs/pull/663#issuecomment-318627609.

Changed the EIP to explicitly list both options (in #2038).

**Comment by axic** [comment]
Added discussion URL: https://ethereum-magicians.org/t/eip-663-unlimited-swap-and-dup-instructions/3346

**Pull request merged**